apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -65,20 +66,26 @@ start_link(Protocol, Request, Fun) ->
 
 -spec(init(pid(), protocol(), binary(), reply_fun()) -> ok).
 init(Parent, Protocol, Request, Fun) ->
+    StartTime = erlang:monotonic_time(millisecond),
     proc_lib:init_ack(Parent, {ok, self()}),
     DNSMessage = #dns_message{} = dns:decode_message(Request),
     Questions = DNSMessage#dns_message.questions,
-    case dcos_dns_router:upstreams_from_questions(Questions) of
+    {Upstreams, Zone} = dcos_dns_router:upstreams_from_questions(Questions),
+    case Upstreams of
         [] ->
             report_status([?APP, no_upstreams_available]),
             Response = failed_msg(Protocol, DNSMessage),
-            ok = reply(Fun, Response);
+            ok = reply(Fun, Response),
+            prometheus_counter:inc(dns_forwarder_failures_total, [Zone, no_upstream]);
         internal ->
             Response = internal_resolve(Protocol, DNSMessage),
-            ok = reply(Fun, Response);
-        Upstreams ->
+            ok = reply(Fun, Response),
+            report_query_response_time(StartTime, Zone);
+        _ ->
             Upstreams0 = take_upstreams(Upstreams),
-            resolve(Protocol, Upstreams0, Request, Fun)
+            FailMsg = failed_msg(Protocol, DNSMessage),
+            external_resolve(Protocol, Upstreams0, Request, Fun, FailMsg, Zone),
+            report_query_response_time(StartTime, Zone)
```

Comment:
Imagine there are two upstream dns servers, one of them times out, but another one works perfectly fine with sub-ms delay. In that case current metric will show 5000ms/+Inf.